### PR TITLE
Redis nonce

### PR DESCRIPTION
**Pull request by maxneuvians**
Part of #265. Creates a new endpoint that generates a 32 byte nonce and stores it in a Redis cache for 30 seconds. This is part of the device attestation process for android. As seen here

Screen Shot 2020-09-04 at 9 47 24 AM

Source: https://www.netguru.com/codestories/stay-safe-with-safetynet-attestation-api-in-android

### CODE REVIEW

File: pkg/server/event.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+type eventServlet struct {
+	cache persistence.RedisConn
+}
+
+func (s *eventServlet) RegisterRouting(r *mux.Router) {
+	r.HandleFunc("/event/nonce", s.eventNonce)
+}
+
+func (s *eventServlet) eventNonce(w http.ResponseWriter, r *http.Request) {
```
Comment (CalvinRodo) [inline pkg/server/event.go:25]: Do we need to handle OPTIONS for this endpoint?

**Review by CalvinRodo — COMMENTED** [review]
Is there any requirement to validate the nonce on the server side, or is that coming in a later PR?

File: pkg/persistence/redis.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+func DialRedis(opt *redis.Options) RedisConn {
+	rdb := redis.NewClient(opt)
+	return &redisConn{rdb: rdb}
+}
+
+func (c *redisConn) GenerateNonce() (string, error) {
+	nonce := make([]byte, 24)
+	for tries := 5; tries > 0; tries-- {
```
[the remark this answers is not in the document]
Comment (maxneuvians) [reply pkg/persistence/redis.go:29]: Good point, will do!

File: pkg/persistence/redis.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+		_, err := rand.Read(nonce)
+
+		if err != nil {
+			return "", err
+		}
+
+		encodedNonce := base64.StdEncoding.EncodeToString(nonce)
+
+		err = c.rdb.Set(encodedNonce, true, time.Second*30).Err()
```
[the remark this answers is not in the document]
Comment (maxneuvians) [reply pkg/persistence/redis.go:39]: Good point, will do!

File: pkg/server/event.go (hunk 1 above)
Comment (maxneuvians) [reply pkg/server/event.go:25]: Depends on if the browser does the call or something else in the OS.

**Comment by maxneuvians** [comment]
Can test functionality: `curl -X POST https://covid-alert-redis-nonce-uqrn5e.herokuapp.com/event/nonce`

**Review by CalvinRodo — APPROVED** [review]
LGTM!

**Comment by maxneuvians** [comment]
Closing until this feature comes up again

**Pull request closed without merging**